Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,190 @@
 
 import NIOConcurrencyHelpers
 
+struct SocketChannelLifecycleManager {
+    // MARK: Types
+    enum State {
+        case neverRegistered
+        case registeredNeverActivated
+        case registered
```

Comment:
Hrm, do we want allow a channel to go registered -> active -> registered -> active? I'd be inclined to want `Channel`s to be single-use constructs.

Reply:
@Lukasa I think you are right... it should be single use.

Reply:
agreed